imap_processing/codice/codice_l2.py

```diff
@@ -13,19 +13,211 @@
 from pathlib import Path
 
 import numpy as np
+import pandas as pd
 import xarray as xr
 
+from imap_processing import imap_module_directory
 from imap_processing.cdf.imap_cdf_manager import ImapCdfAttributes
 from imap_processing.cdf.utils import load_cdf
-from imap_processing.codice.constants import HALF_SPIN_LUT
+from imap_processing.codice.constants import (
+    HALF_SPIN_LUT,
+    HI_ELEVATION_ANGLE,
+    HI_GAIN_MODES,
+    HI_SPIN_ANGLE,
+    HI_SSD_ID_TO_INDEX,
+)
 
 logger = logging.getLogger(__name__)
 logger.setLevel(logging.INFO)
 
 
+def codice_hi_direct_event(l1a_data: xr.Dataset) -> xr.Dataset:
+    """
+    Calculate Hi Direct Event physical quantities.
+
+    Convert the following data variables to physical units using
+    calibration data:
+       - ssd_energy
+       - tof
+       - elevation_angle
+       - spin_angle
+    The other data variables require no changes
+    See section 11.1.2 of algorithm document
+
+    Parameters
+    ----------
+    l1a_data : xarray.Dataset
+        The input dataset containing the raw data variables.
+
+    Returns
+    -------
+    data : xarray.Dataset
+        The dataset with the physical quantities calculated.
+    """
+    # ---------------------------------------------------
+    # Carry these data in L2 without any modification
+    #   Gain
+    #   Multi-Flag
+    #   Spin Number
+    # ---------------------------------------------------
+    l2_dataset = xr.Dataset()
+
+    l2_dataset["gain"] = (l1a_data["gain"].dims, l1a_data["gain"].data)
+    l2_dataset["multi_flag"] = (
+        l1a_data["multi_flag"].dims,
+        l1a_data["multi_flag"].data,
+    )
+    l2_dataset["spin_number"] = (
+        l1a_data["spin_number"].dims,
+        l1a_data["spin_number"].data,
+    )
```

Comment:
Can you do this instead?

```suggestion
    l2_dataset = l1a_data[["gain", "multi_flag", "spin_number"]].copy(deep=True)
```